Add score masking to seven atari environments

stewy33 opened the pull request:
Fixes #61 

Added score masking to the seven atari environments from the RLHF paper. I used a black background to cover the score, to cover enemy ship count for BeamRider, and cover the speedometer for Enduro.

Note that the number of lives is unmasked. This matches the original implementation in the RLHF paper. However, it seems that episode boundaries could be inferred from there. What do we think about this choice?

![image](https://user-images.githubusercontent.com/3607129/194654826-04fad12f-9723-4f59-aa4a-0579bbee374c.png)

AdamGleave commented:
@Rocamonde, do you mind reviewing this PR?

> However, it seems that episode boundaries could be inferred from there. What do we think about this choice?

It's an interesting question. I'd guess number of lives is often decision relevant for the agent -- e.g. it wants to be more risk averse when only one life left. So I lean against masking it, though I agree it introduces a confounder. Having an option to mask it (that we could leave off by default) is probably the best thing to do. But, masking the score is a definitive improvement over not masking it so this shouldn't hold up the PR.

Rocamonde commented on tests/test_envs.py:
```diff
@@ -56,6 +56,17 @@ def test_atari_space_invaders():
+            ATARI_ENVS,
+        ),
+    )
+    assert len(non_masked_environments) == 0
```
So I guess now we're forcing all environments to be masked? Is this what we want? I don't have enough information to have a strong opinion on this, but intuitively it seems useful that a library providing algorithm-agnostic environment implementations should not enforce design choices only relevant to certain algorithms.

Rocamonde replied:
Unless I'm missing something important, I would probably favor having both the masked and unmasked environments. The test you could add then is make sure there is always one of each, even though this would force anyone adding a new environment to also add the masks, but this is probably fine?

AdamGleave replied:
Agree we want to still support the non-masked variants.

Rocamonde reviewed (changes requested):
Thanks so much for taking the time to submit this PR! Overall I agree with the design choices and implementation. There are some major comments that I would like to see addressed:

- Masking `obs_from_state` instead of environment interaction methods
- Whether we want to force all registered environments to be masked (even if users could still manually registered non-masked environments)
- Whether we should have wrappers always inherit from our base class for type safety.

By default, I would favor the solutions I left in the comments, but happy to hear your thoughts on them. Since these are fairly major design choices for the project, maybe @AdamGleave wants to weigh in on these too?

stewy33 commented on src/seals/util.py:
```diff
@@ -23,6 +23,47 @@ def step(self, action):
+                Can support RGB colors by being a sequence of values [r, g, b].
+        """
+        super().__init__(env)
+        self.fill_value = np.array(fill_value, env.observation_space.dtype)
```
So the way the code is written, either works for RGB images, but I can add a test for it. If you provide one value, it will fill all channels with that value and if you provide a sequence it will fill the channels with the sequence. I did this because I wanted people to be able to mask with non-monochrome colors if they wanted to.

Rocamonde replied:
Okay, so you're just relying on broadcasting, and assuming that the user will pass a sensible input that numpy will be able to broadcast. That seems fine

AdamGleave commented on src/seals/util.py:
```diff
@@ -23,6 +23,47 @@ def step(self, action):
         return obs, rew, False, info
 
 
+class MaskScoreWrapper(gym.Wrapper):
```
Hm -- I don't think any of the Atari environments implement `ResettablePOMDP` though? So not convinced we need to fix it in this PR.

I do agree this is a problem with wrappers in general, but that seems out of scope for this PR.

Rocamonde replied:
Sounds good, I'll make an issue for this.

AdamGleave commented:
> * Masking `obs_from_state` instead of environment interaction methods
> * Whether we should have wrappers always inherit from our base class for type safety.

These suggestions seem to be based on the assumption that the Atari environments implement the `ResettablePOMDP` interface, but they don't -- they're base `gym.Env` implementations. There's no `obs_from_state` method for us to mask.

I guess we could make the state be the non-masked observation, and mask it in the observation? That could work but I don't see the benefit.

stewy33 commented on src/seals/util.py:
```diff
@@ -23,6 +23,47 @@ def step(self, action):
+    def step(self, action):
+        """Returns (obs, rew, done, info) with masked obs."""
+        obs, rew, done, info = self.env.step(action)
+        return np.where(self.mask, obs, self.fill_value), rew, done, info
```
Done

stewy33 commented on src/seals/atari.py:
```diff
@@ -1,15 +1,44 @@
 
+SCORE_REGIONS: Dict[str, List[Dict[str, int]]] = {
+    "BeamRider": [
+        dict(x0=5, x1=20, y0=45, y1=120),
```
Done

stewy33 commented:
Hi, I just incorporated some of the suggested changes!

I added the option of having unmasked atari environments. The naming convention is now `seals/BeamRider-v5` for masked environments and `seals/BeamRider-Unmasked-v5` for unmasked environments. The idea for the naming was to suggest that the masked environments are the standard ones. Let me know if you think I should include a warning when loading unmasked environments.

Rocamonde commented on src/seals/util.py:
```diff
@@ -51,17 +52,21 @@ def __init__(
         self.mask = np.ones(env.observation_space.shape, dtype=bool)
         for r in score_regions:
-            self.mask[r["x0"] : r["x1"], r["y0"] : r["y1"]] = 0
+            assert r["x"][0] < r["x"][1] and r["y"][0] < r["y"][1]
```
Since this is a public method (users could create their own wrapper beyond our internal usage for seals-defined environments) you should raise a ValueError instead (and add a corresponding test). Passing in the wrong values is very much a possibility. 

`assert` is to be used when something should always behave in a certain way by virtue of the purported logic of the program. This allows catching logical bugs and reassuring code checkers. This would be fine if only our own internally defined masks could ever be used. However, when something is part of the public API and therefore contingent on user input, we cannot really _assert_ that a user won't pass the wrong value. See https://stackoverflow.com/questions/17530627/python-assertion-style#:~:text=The%20assert%20statement%20should%20only,user%20input%20or%20the%20environment. and https://wiki.python.org/moin/UsingAssertionsEffectively

What you _can_ do, however, is have tests that assert that our internally defined masks verify this. That, plus a `with raises` test on the MaskScoreWrapper API should be enough to thoroughly test this.

stewy33 replied:
Thanks for the explanation!

Rocamonde commented on src/seals/util.py:
```diff
@@ -51,17 +52,21 @@ def __init__(
+            assert r["x"][0] < r["x"][1] and r["y"][0] < r["y"][1]
+            self.mask[r["x"][0] : r["x"][1], r["y"][0] : r["y"][1]] = 0
+
+    def _mask_obs(self, obs):
```
Nice! Thanks for adding this. Code looks cleaner now IMO.

Rocamonde commented on tests/test_envs.py:
```diff
@@ -26,7 +26,11 @@
 
 ATARI_ENVS: List[str] = [
-    _seals_name(gym_spec) for gym_spec in seals.GYM_ATARI_ENV_SPECS
+    _seals_name(gym_spec, masked=False) for gym_spec in seals.GYM_ATARI_ENV_SPECS
```
Maybe define the masked/unmasked envs as separate constants? and then combine them. Probably more readable and it would also prevent black from breaking the lines in strange ways.

stewy33 replied:
Good idea! Done.

stewy33 commented on src/seals/atari.py:
```diff
@@ -59,29 +63,39 @@ def _supported_atari_env(gym_spec: gym.envs.registration.EnvSpec) -> bool:
+    name = "seals/" + slash_separated[-1]
+
+    if not masked:
+        last_hyphen_idx = name.rfind("-")
```
Well, in the `_supported_atari_env` method, we already only support an Atari environment if it ends with "-v4" or "-v5". So I think this is ok for now.

stewy33 commented on src/seals/atari.py:
```diff
@@ -1,44 +1,48 @@
+        if score_region is None:
+            raise ValueError(
+                "Requested environment does not yet support masking. "
+                + "See https://github.com/HumanCompatibleAI/seals/issues/61.",
```
Thanks, done.

Rocamonde reviewed (approved):
LGTM, only a minor fix on the test cases. Please fix before merging. The tests say pending, you might want to re-trigger the pipeline.

AdamGleave commented:
> 

@Rocamonde stewy33 doesn't have permission to trigger them (our tests don't run on fork), I have done this now with `git-push-fork-to-upstream-branch origin stewy33:master`

stewy33 commented:
Hi, I'm new to contributing to open-source projects, so I'm wondering if
1. I need to manually re-trigger the tests
2. Whether you guys prefer me to rebase or merge my changes and whether I should squash my commits (and if so, how much of this is automated).

Rocamonde commented:
> @Rocamonde stewy33 doesn't have permission to trigger them (our tests don't run on fork), I have done this now with `git-push-fork-to-upstream-branch origin stewy33:master`

Thanks, just re-triggered this after the most recent changes.

@AdamGleave what's gonna happen when we do this for multiple open PRs? (apparently the way it works is by pushing to the `trigger-integration` branch.)

@stewy33:

> 1. I need to manually re-trigger the tests

Don't worry, I've just taken care of that. Normally it's automatic, but since your PR is from a fork, it's not.



> 2\. Whether you guys prefer me to rebase or merge my changes and whether I should squash my commits (and if so, how much of this is automated).

IIRC the desirable option is the only one enabled when you press the merge button on the PR. AFAIK we always squash and merge (Adam correct me if this is wrong). Just click the button on the PR once the tests pass and GitHub should automatically do this and close the PR.

Rocamonde commented:
Seems like you have failing tests due to two small mistakes. 

`test_atari_unmasked_env_naming` is failing because you mixed up two ways to check this and forgot to decide on one. SO currently you're doing
```python
        noncompliant_envs = [
            (_get_score_region(name) is None and "Unmasked" not in name)
            for name in ATARI_ENVS
        ]
        assert len(noncompliant_envs) == 0
```
what you actually get is a list of bools with all elements False. So you can either
```python
        noncompliant_envs = [
            name
            if (_get_score_region(name) is None and "Unmasked" not in name)
            for name in ATARI_ENVS
        ]
```
or
```python
        is_each_env_noncompliant = [
            (_get_score_region(name) is None and "Unmasked" not in name)
            for name in ATARI_ENVS
        ]
        assert not any(is_each_env_noncompliant)
```
I probably prefer the first TBH.

Then for the second error, we manually hardcoded the names of some envs not to check because they take a while to show determinism. But since now we also have unmasked and masked versions, your unmasked envs are getting checked again and the tests are failing.
```python
            # these environments take a while for their non-determinism to show.
            slow_random_envs = [
                "seals/Bowling-v5",
                "seals/Frogger-v5",
                "seals/KingKong-v5",
                "seals/Koolaid-v5",
                "seals/NameThisGame-v5",
            ]
```

I would just manually add the other tests to this list for now. Once you're at it, check that we have no hardcoded lists of environments anywhere else that are not being updated.

AdamGleave commented:
> I would just manually add the other tests to this list for now. 

Could also change the test to check if the prefix of the env name matches and skip accordingly.

stewy33 commented:
Hi, I added another test to hopefully past code coverage. Could one of you re-trigger tests? If they pass, I think we're ready to merge.

Rocamonde commented:
@AdamGleave codecov is being annoying, I think it thinks the CI failed and is not reporting coverage, but I checked on the website and all looks good. Can you override this? the PR LGTM.

AdamGleave commented on src/seals/atari.py:
```diff
@@ -1,15 +1,48 @@
-from seals.util import AutoResetWrapper, get_gym_max_episode_steps
+from seals.util import AutoResetWrapper, MaskScoreWrapper, get_gym_max_episode_steps
 
+SCORE_REGIONS: Dict[str, List[Dict[str, Tuple[int, int]]]] = {
```
You use `List[Dict[str, Tuple[int, int]]]` in three places in your code -- consider defining it as a type? Like:

```python
MaskedRegionSpecifier = List[Dict[str, Tuple[int, int]]]
```

I'd also consider using a named tuple instead of `dict` to enforce that `x` and `y` are both present.

AdamGleave commented on src/seals/atari.py:
```diff
@@ -1,15 +1,48 @@
+        score_region = _get_score_region(atari_env_id)
+        if score_region is None:
+            raise ValueError(
+                "Requested environment does not yet support masking. "
```
Nice, thanks for adding the informative error message :)

AdamGleave commented on src/seals/util.py:
```diff
@@ -23,6 +23,56 @@ def step(self, action):
+    def __init__(
+        self,
+        env: gym.Env,
+        score_regions: List[Dict[str, Tuple[int, int]]],
```
(If you did define a type alias this file would be the natural place to do it.)

AdamGleave commented on src/seals/util.py:
```diff
@@ -23,6 +23,56 @@ def step(self, action):
+class MaskScoreWrapper(gym.Wrapper):
+    """Mask a list of box-shaped regions in the observation to hide reward info.
+
+    Intended for environments whose observations are raw pixels (like atari
```
```suggestion
    Intended for environments whose observations are raw pixels (like Atari
```

AdamGleave commented on src/seals/util.py:
```diff
@@ -23,6 +23,56 @@ def step(self, action):
+        self.mask = np.ones(env.observation_space.shape, dtype=bool)
+        for r in score_regions:
+            if r["x"][0] >= r["x"][1] or r["y"][0] >= r["y"][1]:
+                raise ValueError('Invalid region: "x" and "y" must be increasing.')
```
Nice input validation!

AdamGleave commented on tests/test_util.py:
```diff
@@ -2,9 +2,21 @@
+from seals import GYM_ATARI_ENV_SPECS, util
+
+
+def test_mask_score_wrapper_enforces_spec():
```
It might be nice to add a test that actually checks `MaskScoreWrapper` masks the observations -- e.g. you could have a dummy environment that returns all-ones, a dummy mask config, and then just check that region (and only that region) is zero.

I won't insist on it though, the `MaskScoreWrapper` implementation is simple and readable already so is unlikely to have a bug, and you've already done a lot of work in this PR!

AdamGleave reviewed (commented):
@Rocamonde codecov seems to have fixed itself unless you did something?

@stewy33 thanks for all your work on this PR. Got around to reviewing it. Pretty much looks good, a couple of minor suggestions. After that should be ready to merge :)

stewy33 commented:
Hi, could someone trigger the tests one last time before merge? Just incorporated the desired minor changes.

AdamGleave reviewed (approved):
LGTM. Thanks for your patience with this! Also I've changed the CI set up so tests should hopefully run OK on forked repos in futuer.
